safe_control_gym/controllers/mpc/mpc.py

```diff
@@ -462,3 +462,6 @@ def run(self,
             raise Exception("[ERROR] mpc.run().py: MPC could not find a solution for the first step given the initial conditions. "
                   "Check to make sure initial conditions are feasible.")
         return deepcopy(self.results_dict)
+
+    def reset_before_run(self, obs, info=None, env=None):
```

Comment:
hmmmm. I guess everything in the mpc reset is what needs to be reset before a run IMO. Like making sure you have the right reference and dynamics set. I'm not sure what you mean by 'Also, shouldn't the base reset function reset the local env?'

Reply:
Honestly this was so long ago I barely remember, but I was saying shouldn't ctrl.reset() also do ctrl.env.reset()? And since reset is the intense version, shouldnt reset call reset_before_run, not the other way around? idk if that makes sense

Reply:
All good. No i don't think so. You initialize the controller with an env that has a prior model, and you don't want to change any of that on the controller reset. The experiment scripts classes will reset the env for you, and when you do that, it returns an observation and info dict. I don't know what the controller would do with this information and thus I don't know why the controller env should be reset here.

Reply:
Hmmm maybe creating the Experiment class made all these lines unnecessary, but if you look you'll find that every other controller resets the environment on reset (the RL ones do it only while training it seems). I guess maybe it's no longer useful.

Reply:
Ya, I guess practically, I don't know if resetting the environment for MPC would help at all. The env you actually run with will just get reset when the env needs to get reset.

Reply:
Hmmm question: lets say the env given to the controller has randomized priors. when the controller is reset, would we want to reset the environment, get new priors, and use those new priors (i.e. essentially a whole new experiment)? Idk if that makes sense. Trying to think if there is any reason to reset the env being used in the controller

Reply:
hmmmm good question. This situation hasn't really come up yet as we now assume the prior parameters are defined through the controller. So, even if you are resetting the environments with new mass and inertias, the controller is still using whatever prior it started out with, which I think is more the intended use case. Additionally, neither the environments or the controllers have a way to randomize the prior model. 

We also rarely run with the controller environment. We almost always run with a different instantiation of the environment that is now passed to the experiment class. So one situation that I now see is if the experiment class environment is randomizing between resets, and you want to update the controller with that environment's mass properties, there isn't a smooth way to do this. However, I don't really see this being useful, because part of the repeatability is being able to define the prior through the YAML. Like I could see having one prior, and then seeing how it performs with on systems with randomized inertial properties. Would the MPSC need this ability?

Reply:
I think you're right, sounds good. We should update the other controllers to also not reset their environment. Tbh I think ideally the controllers wouldn't take in an environment, they would take in like all the variables of the environment and use that data, but very few controllers need a fully functional environment inside of them. thats a whole other thing tho